modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java

```diff
@@ -368,30 +402,52 @@ protected void initChannel(Channel ch) throws Exception {
             if (transport.readTimeoutMillis > 0) {
                 ch.pipeline().addLast("read_timeout", new ReadTimeoutHandler(transport.readTimeoutMillis, TimeUnit.MILLISECONDS));
             }
-            final HttpRequestDecoder decoder = new HttpRequestDecoder(
-                handlingSettings.maxInitialLineLength(),
-                handlingSettings.maxHeaderSize(),
-                handlingSettings.maxChunkSize()
-            );
+            final HttpRequestDecoder decoder;
+            if (headersAuthenticator != null) {
+                decoder = new HttpRequestDecoder(
+                    handlingSettings.maxInitialLineLength(),
+                    handlingSettings.maxHeaderSize(),
+                    handlingSettings.maxChunkSize()
+                ) {
+                    @Override
+                    protected HttpMessage createMessage(String[] initialLine) throws Exception {
+                        DefaultHttpRequest httpRequest = (DefaultHttpRequest) super.createMessage(initialLine);
+                        return headersAuthenticator.wrapNewMessage(httpRequest);
+                    }
+
+                    @Override
+                    protected HttpMessage createInvalidMessage() {
+                        DefaultHttpRequest httpRequest = (DefaultHttpRequest) super.createInvalidMessage();
+                        return headersAuthenticator.wrapNewMessage(httpRequest);
+                    }
+                };
+            } else {
+                decoder = new HttpRequestDecoder(
+                    handlingSettings.maxInitialLineLength(),
+                    handlingSettings.maxHeaderSize(),
+                    handlingSettings.maxChunkSize()
+                );
+            }
             decoder.setCumulator(ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+            ch.pipeline().addLast("decoder", decoder);
+            if (headersAuthenticator != null) {
+                Netty4HttpHeaderValidator validator = new Netty4HttpHeaderValidator(headersAuthenticator::authenticateMessage);
+                ch.pipeline().addLast("header_validator", validator);
+            }
```

Comment:
This is the second and last change to the pipeline. This is the channel handler that runs the authentication.